[Hub Generated] Review request for Microsoft.RecoveryServices to add version 2018-07-10

sriramvu opened the pull request:
If you are a MSFT employee you can view your [work branch via this link](https://portal.azure-devex-tools.com/app/branch/sriramvu/azure-rest-api-specs/dev-recoveryservicessiterecovery-Microsoft.RecoveryServices-2018-07-10).

### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

sriramvu committed:
Adds base for updating Microsoft.RecoveryServices from version stable/2018-01-10 to version 2018-07-10

sriramvu committed:
Updates readme

sriramvu committed:
Updates API version in new specs and examples

sriramvu committed:
Changes with new API version 2018-07-10

AutorestCI commented:
# Automation for azure-sdk-for-python

AutorestCI commented:
# Automation for azure-sdk-for-ruby
A PR has been created for you:
https://github.com/Azure/azure-sdk-for-ruby/pull/2236

AutorestCI commented:
# Automation for azure-sdk-for-go

AutorestCI commented:
# Automation for azure-sdk-for-java
Nothing to generate for azure-sdk-for-java

azuresdkci commented:
Can one of the admins verify this patch?

AutorestCI commented:
# Automation for azure-sdk-for-node
A PR has been created for you:
https://github.com/Azure/azure-sdk-for-node/pull/4688

AutorestCI commented:
# Automation for azure-sdk-for-js

sriramvu commented:
Introduced new API version 2018-07-10 for Azure Site Recovery (ASR) service:

Breaking changes:
* modified properties in InMageAzureV2EnableProtectionInput and InMageAzureV2ReplicationDetails using new contracts InMageAzureV2DiskInputDetails and InMageAzureV2ManagedDiskDetails
* removed roleAssignment

Incremental changes:
* new APIs addDisks, removeDisks, resolveHealthErrors, replicationSupportedOperatingSystems got added

Ran _autorest --azure-validator_ and sanity tools under _OpenAPI Validation Tools_ and found no issues.

nschonni commented on specification/recoveryservicessiterecovery/resource-manager/Microsoft.RecoveryServices/stable/2018-07-10/service.json:
```diff
@@ -0,0 +1,17459 @@
+            "type": "string"
+          },
+          {
+            "name": "migrationI
```
Resolvability -> Resolvable? Could just be  a boolean that way

innosam replied:
Basically, we thought of the boolean approach. But decided against it, due to the limitation in future extendability if we ever wanted a new value apart from the current set i.e Allowed, NotAllowed.

ravbhatnagar commented on specification/recoveryservicessiterecovery/resource-manager/Microsoft.RecoveryServices/stable/2018-07-10/service.json:
```diff
@@ -3322,6 +3476,83 @@
         }
       }
     },
+    "/Subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RecoveryServices/vaults/{resourceName}/replicationFabrics/{fabricName}/replicationProtectionContainers/{protectionContainerName}/replicationProtectedItems/{replicatedProtectedItemName}/ResolveHealthErrors": {
```
nit: camel case for subscriptions

sriramvu replied:
We will be taking later as it reflects all the existing APIs too.

ravbhatnagar commented on specification/recoveryservicessiterecovery/resource-manager/Microsoft.RecoveryServices/stable/2018-07-10/service.json:
```diff
@@ -3180,6 +3257,83 @@
         }
       }
     },
+    "/Subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RecoveryServices/vaults/{resourceName}/replicationFabrics/{fabricName}/replicationProtectionContainers/{protectionContainerName}/replicationProtectedItems/{replicatedProtectedItemName}/removeDisks": {
```
ensure linked access check is happening in ARM manifest.

sriramvu replied:
As @innosam has communicated over mail, we are updating manifest in parallel.

ravbhatnagar commented on specification/recoveryservicessiterecovery/resource-manager/Microsoft.RecoveryServices/stable/2018-07-10/service.json:
```diff
@@ -7189,6 +7485,17 @@
           "description": "The data pending at source virtual machine in MB.",
           "type": "number"
         },
+        "diskState": {
```
this should be modeled as enum

innosam replied:
As far as I know using enums is not considered a good practice. Are you are referring allowed values construct?

ravbhatnagar replied:
why is enums not a good practice? you can set modelasString:true and that will not be a breaking change when you add new values.

ravbhatnagar commented on specification/recoveryservicessiterecovery/resource-manager/Microsoft.RecoveryServices/stable/2018-07-10/service.json:
```diff
@@ -8127,13 +8502,6 @@
             "type": "string"
           }
         },
-        "roleAssignments": {
```
why is this property being removed?

sriramvu replied:
Om Nishant <omnishan@microsoft.com>: After the roles and responsibilities were revisited to keep up with compliance and customer expectations, extraneous fields were removed to prevent any misleading and unrealistic expectations.

ravbhatnagar replied:
removing properties from an existing API-versions is a breaking change. It should be done in a new api-version

ravbhatnagar replied:
what happens when customer sets this property using an old api-version which supported it and gets the resource using new api-version. And then may be after that, does a PUT again from the new api-version

sriramvu replied:
roleAssignments under AzureToAzureVmSyncedConfigDetails is present only in response of GET ReplicationProtectedItem.

Actually these are role assignments of the Virtual Machine that is getting replicated (Virtual Machine and ReplicationProtectedItem have 1:1 association in ASR).
We were storing & syncing roleAssignments of the Virtual Machine - which we have stopped doing (shared the reason from Om Nishant in previous comment).

So, there is no PUT and only GET associated with the property which we have removed.

ravbhatnagar commented on specification/recoveryservicessiterecovery/resource-manager/Microsoft.RecoveryServices/stable/2018-07-10/examples/SupportedOperatingSystems_Get.json:
```diff
@@ -0,0 +1,20 @@
+        "type": "Microsoft.RecoveryServices/vaults/replicationSupportedOperatingSystems",
+        "id": "/Subscriptions/bc403605-c2b0-43dd-abe9-0162124b1ee1/resourceGroups/oneBoxRG/providers/Microsoft.RecoveryServices/vaults/oneBoxRSVault/replicationSupportedOperatingSystems/Default",
+        "properties": {
+	  "supportedOsList":[]
```
complete the example

sriramvu replied:
Completed.

ravbhatnagar reviewed (commented):
few comments

ravbhatnagar commented:
There is an open issue with this design which is due to a limitation in ASR service and second this has already been implemented on service side. Ideally addDisk, removeDisk, resolveHealthError kind of scenarios should have been addressed via a PUT or PATCH on the parent resource instead of as a POST action. Please make sure you are doing the right auth check in the manifest  
Pasting email conversation below for some context - 

_Gaurav Bhatnagar
Thu 11/15/2018 12:39 AM
Sure, please let me know if you have questions on Linked access checks. 
Please note this is not an ideal API design. More so, when it seems like this path was chosen to overcome engineering complexities/challenges. We strongly discourage the use of POST to change resource models like in this case. Imagine if every service in Azure starts enabling POST APIs to update certain properties on the resources it owns. Soon the API surface in Azure will become confusing, inconsistent and hard to use. But since this is already implemented/deployed on the service side, we cant change it now. 
Apologies again for missing this thread earlier. 
Thanks,
Gaurav


Sent from Outlook
Harkirat Singh
Tue 11/13/2018 10:38 PM
Please find the comments inline.

From: Gaurav Bhatnagar <Gaurav.Bhatnagar@microsoft.com> 
Sent: 14 November 2018 02:42
To: Harkirat Singh <hsingh@microsoft.com>
Cc: Himanshu Saini <Himanshu.Saini@microsoft.com>; Mukesh Kumar (CDAP) <mukeshk@microsoft.com>; Vijay Narayan <vijay.narayan@microsoft.com>; Azure Resource Manager RP API Review <armapireview@microsoft.com>
Subject: Re: SiteRecovery API to support add/remove protected disks
Ok, both these are not ideal. Wanted to clarify a couple of things -

For #1 comment below - Does the second PUT fail on an existing resource fail if some property values are changed in the payload. If a property is patchable using a PATCH API, it should be updatable using a PUT as well. 

<Harkirat> I can confirm that for replication protected item, the second PUT call with the same input is handled idempotence. While, if some of the input  is changed, we fail with a user error.

 

#2a -Seems like an internal implementation detail.

<Harkirat> yes this is specific to asr, but the jobs are available as ARM resource replicationJobs/{jobname}.

And, when the get is done on the resource we expose the list of task executed for a particular long running job.

 

#2b - IMO, it doesn't complicate the scenario. May be it complicates the implementation on the backend. But, as far as user experience is concerned, it provides flexibility to the user. The user can choose to update any number of properties as part of the patch. the user doesn't have to update a few properties using PATCH and somehow figure out that for disk update they need a separate POST API. 

<Harkirat> Yes, it does complicate the implementation. From user experience perspective, the bright side for the POST call is that user need not provided the existing disk set and their parameters, when adding a new disk.

 

Moreover, in the current model, RBAC is broken. Since now someone without /write permissions on replicationProtectedItems can change its state by doing a POST.

It looks like this has already been implemented. Can you please confirm? If yes, I would recommend that we atleast do a linked access check to ensure the RBAC is not an issue.

<Harkirat> Yes, we have already completed the implementation.  I will definitely ensure that the linked access check to ensure the RBAC is not an issue.

Thanks for bringing this point, I will reach you if I have doubts on “how to” of this required action.

 

Thanks,

Gaurav

 

Sent from Outlook

From: Harkirat Singh
Sent: Tuesday, November 13, 2018 1:39 AM
To: Gaurav Bhatnagar
Cc: Himanshu Saini; Mukesh Kumar (CDAP); Vijay Narayan
Subject: RE: SiteRecovery API to support add/remove protected disks

 

Hey Gaurav,

 

I was away last week for Diwali holidays, so could not respond.

 

Yes, we did consider the PATCH/PUT call, and eventually decided on separate POST due to following reasons:

PUT
The site recovery service was not designed to have PUT calls update the parameters.

                So, currently none of our PUT calls support update.

 

PATCH
We do have a PATCH call on the protected item, where multiple parameters can be updated.

                But using this verb for add/remove disks have following cons for us:

a.       The PATCH call fires a workflow/job which is generic in nature.

The add/remove disk scenario needed descriptive task which are different from the current PATCH.

 

b.       And, if we do allow the disk update through PATCH, user can club different updates together with add/remove disk, which complicates the scenario.

 

In summary separate API’s is more convenient to the user as only the diff of the disk needs to be provided.

 

Thanks

Harkirat Singh_

sriramvu committed:
Completed the example for SupportedOperatingSystems_Get

ravbhatnagar commented:
@sriramvu - just one pending open question

viverm commented on specification/recoveryservicessiterecovery/resource-manager/Microsoft.RecoveryServices/stable/2018-07-10/service.json:
```diff
@@ -7368,6 +7686,32 @@
+        }
+      ],
+      "properties": {
+        "vmDisksUris": {
```
is naming fine DisksUris?

ravbhatnagar replied:
property names should also be camel cased.

sriramvu replied:
vmDisksUris is camel cased, did not get the comment.
Suggesting a change?

ravbhatnagar replied:
yes, this is correct. I thought you asked a question and I replied saying es, propertynames should also be camel cased.

viverm reviewed (commented):
Service is already out with changes.
No point in holding things for this minor things.

ravbhatnagar commented:
Signing off from ARM side.
